server/src/main/java/org/opensearch/action/admin/cluster/remotestore/stats/TransportRemoteStoreStatsAction.java

```diff
@@ -0,0 +1,159 @@
+                if (Arrays.asList(request.shards()).contains(Integer.toString(shardRouting.shardId().id()))) {
+                    newShardRoutings.add(shardRouting);
+                }
+            });
+        } else {
+            newShardRoutings.addAll(clusterState.routingTable().allShards(concreteIndices).getShardRoutings());
+        }
+        return new PlainShardsIterator(
+            newShardRoutings.stream()
+                .filter(shardRouting -> remoteRefreshSegmentPressureService.getRemoteRefreshSegmentTracker(shardRouting.shardId()) != null)
+                .filter(
+                    shardRouting -> !request.local()
+                        || (shardRouting.currentNodeId() == null
+                            || shardRouting.currentNodeId().equals(clusterState.getNodes().getLocalNodeId()))
+                )
+                .filter(ShardRouting::primary)
+                .filter(shardRouting -> indicesService.indexService(shardRouting.index()).getIndexSettings().isRemoteStoreEnabled())
+                .collect(Collectors.toList())
+        );
+    }
+
+    @Override
+    protected ClusterBlockException checkGlobalBlock(ClusterState state, RemoteStoreStatsRequest request) {
+        return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_READ);
+    }
+
+    @Override
+    protected ClusterBlockException checkRequestBlock(ClusterState state, RemoteStoreStatsRequest request, String[] concreteIndices) {
+        return state.blocks().indicesBlockedException(ClusterBlockLevel.METADATA_READ, concreteIndices);
+    }
+
+    @Override
+    protected RemoteStoreStats readShardResult(StreamInput in) throws IOException {
+        return new RemoteStoreStats(in);
+    }
+
+    @Override
+    protected RemoteStoreStatsResponse newResponse(
+        RemoteStoreStatsRequest request,
+        int totalShards,
+        int successfulShards,
+        int failedShards,
+        List<RemoteStoreStats> responses,
+        List<DefaultShardOperationFailedException> shardFailures,
+        ClusterState clusterState
+    ) {
+        return new RemoteStoreStatsResponse(
+            responses.toArray(new RemoteStoreStats[0]),
+            totalShards,
+            successfulShards,
+            failedShards,
+            shardFailures
+        );
+    }
+
+    @Override
+    protected RemoteStoreStatsRequest readRequestFrom(StreamInput in) throws IOException {
+        return new RemoteStoreStatsRequest(in);
+    }
+
+    @Override
+    protected RemoteStoreStats shardOperation(RemoteStoreStatsRequest request, ShardRouting shardRouting) {
+        IndexService indexService = indicesService.indexServiceSafe(shardRouting.shardId().getIndex());
+        IndexShard indexShard = indexService.getShard(shardRouting.shardId().id());
+        // if we don't have the routing entry yet, we need it stats wise, we treat it as if the shard is not ready yet
+        if (indexShard.routingEntry() == null) {
```

Comment:
shouldn't we have a similar validation here to validate that the underlying index is remote store enabled or not? Otherwise, it might result into NPE resulting into error response on hitting this api with non remote enabled index. Rather we can give clear info back to client.

Reply:
This method is only called when shards are filtered by `shards` method. But let me confirm once

Reply:
Confirmed. this would only be called after shards are filtered by `shards` method above in this class